### PR TITLE
[Yosys] init

### PR DESCRIPTION
**Pull request by sjkelly**
This adds support for building Yosys, the open source synthesis suite for FPGAs and ASICs.

### CODE REVIEW

File: Y/Yosys/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+# Collection of sources required to complete build
+sources = [
+    GitSource("https://github.com/YosysHQ/yosys.git", "2e421feb0ea526468493ab008f3c72beb12c2bc6")
+]
+
+dependencies = [Dependency("boost_jll"),
+                Dependency("Tcl_jll"),
+                Dependency("Readline_jll"),
+                Dependency("Libffi_jll")]
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:16]: Please see #3024
```suggestion
dependencies = [
    Dependency("boost_jll"; compat="=1.76.0"),
    Dependency("Tcl_jll"),
    Dependency("Readline_jll"),
    Dependency("Libffi_jll"; compat="~3.2.2"),
]
```

File: Y/Yosys/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd yosys
+make -j3; make install PREFIX=$prefix
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:21]:
```suggestion
make -j${nproc}
make install PREFIX=${prefix}
```

File: Y/Yosys/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+platforms = supported_platforms(; experimental=true)
+platforms = expand_cxxstring_abis(platforms)
+
+# The products that we will ensure are always built
+products = Product[
+	ExecutableProduct("yosys", :yosys)
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"10")
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:35]: GCC 10 is a bit too much, especially for a C++ program

File: Y/Yosys/build_tarballs.jl (hunk 1 above)
Comment (giordano) [reply Y/Yosys/build_tarballs.jl:16]: Also, is Tcl a runtime dependency or only build time?

Comment (sjkelly) [reply Y/Yosys/build_tarballs.jl:16]: Good question. Let me check locally.

File: Y/Yosys/build_tarballs.jl (hunk 3 above)
Comment (sjkelly) [reply Y/Yosys/build_tarballs.jl:35]: Should I drop that line all together?

File: Y/Yosys/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# Bash recipe for building across all platforms
+script = raw"""
+cd yosys
+make CONFIG=${CC} -j${nproc}
+make install PREFIX=${prefix}
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = filter!(p -> arch(p) != "armv7l" || arch(p) != "armv6l", supported_platforms(; experimental=true))
```
Comment (sjkelly) [inline Y/Yosys/build_tarballs.jl:29]: &&

File: Y/Yosys/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    Dependency("boost_jll"; compat="=1.76.0"),
+    Dependency("Tcl_jll"),
+    Dependency("Readline_jll"),
+    Dependency("Libffi_jll"; compat="~3.2.2"),
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd yosys
+make CONFIG=${CC} -j${nproc}
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:23]: Umpf, of course that doesn't work
```suggestion
if [[ "${target}" == *-apple-* ]] || [[ "${target}" == *-freebsd* ]]; then
    CONFIG=clang
else
    CONFIG=gcc
fi
make CONFIG=${CONFIG} -j${nproc}
```

File: Y/Yosys/build_tarballs.jl (hunk 4 above)
Comment (giordano) [reply Y/Yosys/build_tarballs.jl:34]:
```suggestion
platforms = filter!(p -> arch(p) ∉ ("armv7l", "armv6l"), supported_platforms(; experimental=true))
```

**Comment by sjkelly** [comment]
hmm the `make CONFIG=${CC}` is not handled by the Yosys makefile. They are expecting `gcc` but `cc` is passed. Should this be patched or is there a mechanism to do this here?

**Comment by giordano** [comment]
[`0d27f38` (#3435)](https://github.com/JuliaPackaging/Yggdrasil/pull/3435/commits/0d27f380b6123c6c108101b816605eeadd8b7944) :slightly_smiling_face:

File: Y/Yosys/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd yosys
+if [[ "${target}" == *-apple-* ]] || [[ "${target}" == *-freebsd* ]]; then
+    CONFIG=clang
+else
+    CONFIG=gcc
+fi
+make CONFIG=${CONFIG} -j${nproc}
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:28]:
```suggestion
make CONFIG=${CONFIG} PREFIX=${prefix} -j${nproc}
```

**Comment by giordano** [comment]
Actually we already have ABC, but of course here they use a fork...

**Comment by sjkelly** [comment]
Was about to say. Should we package YosysABC separate?

**Comment by giordano** [comment]
Would you be OK with skipping ABC at the moment?  At least to see whether it'd succeed without it, they're all choking there

**Comment by sjkelly** [comment]
The diff doesn't look like it should affect, and we can use the in-tree ABC. most (all?) of the yosys ABC changes are related to WASM support: https://github.com/berkeley-abc/abc/compare/master...YosysHQ:yosys-experimental

**Comment by sjkelly** [comment]
Please do whatever experiments you like. I am following along and learning.

File: Y/Yosys/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+sources = [
+    GitSource("https://github.com/YosysHQ/yosys.git", "2e421feb0ea526468493ab008f3c72beb12c2bc6")
+]
+
+dependencies = [
+    Dependency("boost_jll"; compat="=1.76.0"),
+    Dependency("Tcl_jll"),
+    Dependency("Readline_jll"),
+    Dependency("Libffi_jll"; compat="~3.2.2"),
+]
```
Comment (sjkelly) [inline Y/Yosys/build_tarballs.jl:18]:
```suggestion
    Dependency("ABC_jll"; compat="~1.1.0"),
]
```

File: Y/Yosys/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd yosys
+if [[ "${target}" == *-apple-* ]] || [[ "${target}" == *-freebsd* ]]; then
+    CONFIG=clang
+else
+    CONFIG=gcc
+fi
+make CONFIG=${CONFIG} PREFIX=${prefix} -j${nproc}
```
Comment (sjkelly) [inline Y/Yosys/build_tarballs.jl:28]:
```suggestion
make CONFIG=${CONFIG} PREFIX=${prefix} LINK_ABC=1 -j${nproc}
```

Comment (giordano) [reply Y/Yosys/build_tarballs.jl:28]: I'd like to try
```suggestion
make CONFIG=${CONFIG} PREFIX=${prefix} ENABLE_ABC=0 -j${nproc}
```
to completely disable ABC, and see what happens, if you're OK

**Comment by sjkelly** [comment]
FYI Nix builds with separate ABC. I cribbed from there: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/yosys/default.nix

**Comment by sjkelly** [comment]
with https://github.com/JuliaPackaging/Yggdrasil/pull/3435/commits/3177fcdef7cd047ef45db1c90c36346e07d746af

I get: 
```
[22:40:16] /opt/i686-linux-gnu/bin/i686-linux-gnu-ld: passes/techmap/abc.o: unrecognized relocation (0x2b) in section `.text'
[22:40:16] /opt/i686-linux-gnu/bin/i686-linux-gnu-ld: final link failed: Bad value
[22:40:16] clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```
on most all platforms. I am setting `preferred_gcc_version=v"8"` per troubleshooting.md.

File: Y/Yosys/build_tarballs.jl [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+make install PREFIX=${prefix}
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = filter!(p -> Sys.isapple(p) || Sys.islinux(p), supported_platforms())
+
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:36]:
```suggestion
platforms = expand_cxxstring_abis(platforms)

```

File: Y/Yosys/build_tarballs.jl [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+fi
+make CONFIG=${CONFIG} PREFIX=${prefix} -j${nproc}
+make install PREFIX=${prefix}
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = filter!(p -> Sys.isapple(p) || Sys.islinux(p), supported_platforms())
+platforms = expand_cxxstring_abis(platforms)
+
```
Comment (giordano) [inline Y/Yosys/build_tarballs.jl:37]:
```suggestion
# For some reasons, building for CXX03 string ABI doesn't actually work, skip it
filter!(x -> cxxstring_abi(x) != "cxx03", platforms)

```

**Pull request merged**